### DISCUSSION TITLE
z transform roadmap discussion

### DISCUSSION THREAD

**Discussion by hanspi42** [sympy/sympy#25040] [Ideas]
@zolabar suggested (and I agree) that we should discuss the roadmap for how to deal with the z transform.

In #24984, @bugmaker2 implemented a general bilinear transform, and the idea of having a class ` TransferFunctionZ` came up.

Meanwhile, in #24985, @edelrosespol is working on the z transform itself.  There one open question is whether the inverse z transform should be a mathematical function containing Dirac impulses or a mathematical function giving weights when evaluated on integers.

My issue #24834 would be precisely for the use with z transforms, as $z^{-1}$ is one delay and $z^{-n}$ for positive integers $n$ are $n$ delays.

@sylee957 and @smichr (as far as I can tell) have done a lot of reviews both on the physics/control parts and on the integrals/transforms parts.

The z transform is special inasmuch as it is a transform between a discrete and a continuous domain. So how should this be implemented in SymPy? Please give me your opinions!

**Comment by sylee957** [comment]
I'm may misunderstood about what do you refer by `functions containing dirac impulses`, and `function giving weights when evaluated on integers`

> whether the inverse z transform should be a mathematical function containing Dirac impulses or a mathematical function giving weights when evaluated on integers.

But I think that it is distinction of `DiracDelta` and some discrete delta. I think that we should follow the notations in standard mathematics as much as possible, especially if we define the rules outselves, unless the other way gives some computational advantage or naturally follows from some algorithms. It is possible to use `KroneckerDelta` or define a new symbolic function if the representation is not convenient, but I don't think that we should just use `DiracDelta` for the reason it looks alike.

**Reply by hanspi42** [reply]
I also would not use it for the $z$ transform, but it is often done: One way to describe sampling mathemathically for a continuous function $f(t)$ is to multiply it by $\sum_i δ(t−iT)$, where $T$ is the sampling time, and $\delta(\cdot)$ is indeed the Dirac delta. If you directly calculate the Fourier transform of $f(t)\sum_i δ(t−iT)$, you get the same solution as when you substitute $z=e^{jωT}$ in the z transform of $f[i]=[f(iT)]$.

I'll write a short theory document explaining it all in the next few days.

**Comment by bugmaker2** [comment]
Just for reference, in Scipy

[`class scipy.signal.lti`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lti.html#scipy.signal.lti) is the continuous LTI system.
It has methods:
`bode([w, n])` | Calculate Bode magnitude and phase data of a continuous-time system.
`freqresp([w, n])` | Calculate the frequency response of a continuous-time system.
`impulse([X0, T, N])` | Return the impulse response of a continuous-time system.
`output(U, T[, X0])` | Return the response of a continuous-time system to input U.
`step([X0, T, N])` | Return the step response of a continuous-time system.
`to_discrete(dt[, method, alpha])` | **Return a discretized version of the current system.**

It can be constructed by [TransferFunction](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.TransferFunction.html#scipy.signal.TransferFunction), [ZeroPolesGain](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.ZerosPolesGain.html#scipy.signal.ZerosPolesGain), and [StateSpace](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.StateSpace.html#scipy.signal.StateSpace)

[`class scipy.signal.dlti`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.dlti.html#scipy.signal.dlti) is the discratized LTI system, which can also be constructed by TransferFunction


In SymPy, we already have the `LTI` and `TransferFunction`. Maybe we can simulate the structure?

---

I planned to work on this as part of my GSoC2023 worklist, as I mentioned in the [brief proposal](https://github.com/sympy/sympy/wiki/GSoC-2023-Application-Baiyuan-Qiu:-Extensions-of-Symbolic-Control-Systems-(sympy.physics.control)).

**Reply by zolabar** [reply]
@bugmaker2, I think, that this roadmap is good, @hanspi42, what do you think?



### to_discrete 

to_discrete(dt[, method, alpha]) | Return a discretized version of the current system.

-> means, a TransferFunction from a future module sympy.physics.control.dlti (TransferFunctionDiscrete? or TransferFunctionZ?) ?

### Filling gaps

freqresp([w, n]) | Calculate the frequency response of a continuous-time system. ->

this can be added to the continuous version, too, as far as I know it is still missing in 

https://docs.sympy.org/latest/modules/physics/control/control_plots.html

### Caution

When translating sympy.physics.control.lti to  sympy.physics.control.dlti be careful, since methods (as stability) may change

### Relation to genuine z-transform from discrete

A possible relation I see between sympy.physics.control.dlti and z-transforms in discrete is, that if an object "difference system" existed, for instance for backward differences:

$$y_k = (x_k-x_{k-1})/T$$

Then applying z-transforms (i.e. from integrals) to it and doeing some alegebra would lead to a corresponding sympy.physics.control.dlti TransferFunction (TransferFunctionDiscrete? or TransferFunctionZ?)

$$y_k = (x_k-x_{k-1})/T$$

$$Y(z) = X(z)(1-z^{-1})/T$$

$$H(z) = \frac{z-1}{T\cdot z}$$

hence something corresponding to

`TransferFunctionDiscrete(z-1, T*z, z)`

Likewise, with the help of sympy.discrete inverse z-transforms, any sympy.physics.control.dlti TransferFunctionDiscrete could be converted to corresponding discrete system.



@hanspi42 , @edelrosespol, @bugmaker2 , @sylee957

**Comment by hanspi42** [comment]
I will have to look at this in detail. Meanwhile I have written up a small document that outlines the problem from the integral-transform point of view. This is related to the questions in #24985 which looks at implementing the z transform as an integraol transform not as a LTI tool, but as a mathematical function.

Look at the Jupyter notebook if you want to see the graphs.

[integral_transforms.pdf](https://github.com/sympy/sympy/files/11213576/integral_transforms.pdf)

[integral_transforms_ipynb.zip](https://github.com/sympy/sympy/files/11213580/integral_transforms_ipynb.zip)

**Comment by hanspi42** [comment]
> Likewise, with the help of sympy.discrete inverse z-transforms, any sympy.physics.control.dlti TransferFunctionDiscrete could be converted to corresponding discrete system.

I think that then proposal of @bugmaker2 and what @zolabar wrote is very good, but I also see a glaring hole: there is no discrete-time function cnstruct yet in SymPy. We should have a way to represent such a thing: a function that only takes on non-zero values for integer arguments. Using the function might activate a generator function or extract a value from a list of argument-value pairs. I think having such a construct would let everything else fall in place quite easily.

What properties and methods should such a function have? Definitely element-wise operations as well as convolution.

What do you think about this?

**Reply by zolabar** [reply]
@hanspi42 , thank you for those deep thougts.

We should have a way to represent such a thing: a function that only takes on non-zero values for integer arguments. Using the function might activate a generator function or extract a value from a list of argument-value pairs.

I have to admit, that I have difficulties to image this. Can we consider a minimal example, maybe with respect of backward differences (some concrete derivatives) or trapezoidal integration? Something for which I know and understand the discrtete transfer function, the corredponding difference equation and its continuous counterparts? Thank you!

**Comment by sylee957** [comment]
The scipy's transfer function API only has the parameter `dt`.
And I think that that parameter can be made symbolic to represent the discrete transfer function
`TransferFunction(num, den, var, dt)`.
But maybe it could be minimal to begin with only `dt=1` case, but I'm not sure if that trivializes things too much.

**Reply by zolabar** [reply]
I also think so.

dt corresponds to „T“ in the coefficient transform functions bilinear, backward_diff, gbt… in SymPy (dev) in control
